packages/runtime/src/templates/edge-shared/utils.ts

```diff
@@ -59,6 +59,34 @@ function isMiddlewareResponse(response: Response | MiddlewareResponse): response
   return 'dataTransforms' in response
 }
 
+// Next 13 supports request header mutations and has the side effect of prepending header values with 'x-middleware-request'
+// as part of invoking NextResponse.next() in the middleware. We need to remove that before sending the response the user
+// as the code that removes it in Next isn't run based on how we handle the middleware
+//
+// Related Next.js code:
+// * https://github.com/vercel/next.js/blob/68d06fe015b28d8f81da52ca107a5f4bd72ab37c/packages/next/server/next-server.ts#L1918-L1928
+// * https://github.com/vercel/next.js/blob/43c9d8940dc42337dd2f7d66aa90e6abf952278e/packages/next/server/web/spec-extension/response.ts#L10-L27
+export function updateModifiedHeaders(response: Response) {
+  const overriddenHeaders = response.headers.get('x-middleware-override-headers') || ''
+
+  if (!overriddenHeaders) {
+    return response
+  }
+
+  const headersToUpdate = overriddenHeaders.split(',')
+
+  for (const header of headersToUpdate) {
+    const oldHeaderKey = 'x-middleware-request-' + header
+    const headerValue = response.headers.get(oldHeaderKey) || ''
+
+    response.headers.set(header, headerValue)
```

Comment:
We don't want to set these in the response - it's the request that needs them, so we should be calling this before calling `context.next()`

Reply:
The `request` headers appear to be fine in that they don't have the `x-middleware-request` string prefixed to every header, nor does it contain the `x-middleware-override-headers` header - it's the `result.response` headers that have them and are not having them cleaned up.

Would be happy to pair on this tomorrow though as there's likely more insight that you have that I'm missing